### PR TITLE
Merge server tests with robustness suite branch

### PR DESCRIPTION
**Pull request by rahulchheda**
Merge Kopia Server/Client changes into `robustness-suite`
Added Fix for the Shippable Pipeline too.

### CODE REVIEW

**Review by redgoat650 — APPROVED** [review]
LGTM. By the way, everything from my end has been pushed upstream, so if you want to open PRs against kopia/kopia master instead that may be helpful. Furthermore, as discussed with @julio-lopez yesterday, we may want to switch the robustness job over to using `master` instead of `robustness-suite` branch, so we'll have one fewer branch to keep synchronized.

**Review by julio-lopez — CHANGES_REQUESTED** [review]
There are a number of linter failures.
Please address those and make sure CI passes.

**Comment by rahulchheda** [comment]
Sorry for the force-push need to rebase this branch to test all the changes, tests running fine.

Added Fix for the Shippable Pipeline too.

**Comment by rahulchheda** [comment]
Should we create a new branch based on `robustness-suite`, and then use that for robustness tests?

**Comment by julio-lopez** [comment]
> Should we create a new branch based on `robustness-suite`, and then use that for robustness tests? cc: @julio-lopez

Once this PR is ready, it can be merged into the `robustness-suite` branch. Then, we can:
- Use the `robustness-suite` branch for the jobs and remove the `robustness-suite-server` one .
- Clean up the `robustness-suite` branch (in terms of commits), to push these changes upstream as well.

ATM, the change for the cert files directory is needed for this PR to pass. @rahulchheda Can you back port that change to this PR?

**Comment by pavannd1** [comment]
@rahulchheda Still failing on the same `os.RemoveAll` line
```
tests/tools/kopiarunner/kopia_snapshotter.go:301:2: defer statements should only be cuddled with expressions on same variable (wsl)
	defer os.RemoveAll(tempDir) // nolint:errcheck
```

File: tests/testenv/cli_test_env.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ func NewCLITest(t *testing.T) *CLITest {
 		t.Name(),
 		"/", "_"), "\\", "_"), ":", "_")
 
-	logsDir := filepath.Join(os.TempDir(), "kopia-logs", cleanName+"."+clock.Now().Local().Format("20060102150405"))
+	logsDir := filepath.Join(t.TempDir(), "kopia-logs", cleanName+"."+clock.Now().Local().Format("20060102150405"))
```
Comment (julio-lopez) [inline tests/testenv/cli_test_env.go:80]: This seems like a spurious change.
Probably it'd be rejected upstream. The change affects many other tests besides the robustness ones.

IIRC, the intention was to be able to preserve the logs for debugging failed tests. I believe by default `t.TempDir()` is deleted even if the test fails.

File: tests/robustness/engine/engine.go [hunk 2]
```diff
@@ -277,3 +299,46 @@ func (e *Engine) init(ctx context.Context) error {
...
+}
+
+// cleanUpServer cleans up the server process.
+func (e *Engine) cleanUpServer() {
+	if e.serverCmd != nil {
+		e.serverCmd.SysProcAttr = &syscall.SysProcAttr{
+			Pdeathsig: syscall.SIGTERM,
+		}
+
+		e.serverCmd.Run() //nolint:errcheck
```
Comment (julio-lopez) [inline tests/robustness/engine/engine.go:342]: What does this do? What is the intent?

File: tests/tools/kopiarunner/kopia_snapshotter.go [hunk 3]
```diff
@@ -227,3 +270,109 @@ func parseManifestListForSnapshotIDs(output string) []string {
...
+}
+
+func certKeyExist(ctx context.Context, tlsCertFile, tlsKeyFile string) error {
+	if err := retry.PeriodicallyNoValue(ctx, retryInterval, retryCount, "waiting for server to start", func() error {
+		if _, err := os.Stat(tlsCertFile); os.IsNotExist(err) {
+			return err
+		}
+		if _, err := os.Stat(tlsKeyFile); os.IsNotExist(err) {
+			return err
+		}
```
Comment (julio-lopez) [inline tests/tools/kopiarunner/kopia_snapshotter.go:371]: IINM, one of the linters will fail here:
```suggestion
		}

		if _, err := os.Stat(tlsKeyFile); os.IsNotExist(err) {
			return err
		}

```

You can run the linter locally with `make lint`

Comment (rahulchheda) [reply tests/tools/kopiarunner/kopia_snapshotter.go:371]: Ran `make lint`, its just fails with `Pdeathsig: syscall.SIGTERM` field not found.

File: tests/robustness/engine/engine.go (hunk 2 above)
Comment (rahulchheda) [reply tests/robustness/engine/engine.go:342]: Cleanup the server process created in background.

File: tests/testenv/cli_test_env.go (hunk 1 above)
Comment (rahulchheda) [reply tests/testenv/cli_test_env.go:80]: Sure, let me fix that.

File: tests/tools/kopiarunner/kopia_snapshotter.go (hunk 3 above)
Comment (julio-lopez) [reply tests/tools/kopiarunner/kopia_snapshotter.go:371]: What OS?

Comment (rahulchheda) [reply tests/tools/kopiarunner/kopia_snapshotter.go:371]: macOS, seems like this code will only work for Linux. What should we do?

File: tests/robustness/engine/engine.go (hunk 2 above)
Comment (julio-lopez) [reply tests/robustness/engine/engine.go:342]: But, what does it do? how does it work?

I may be wrong, but I don't think this works.

File: tests/tools/kopiarunner/kopia_snapshotter.go (hunk 3 above)
Comment (redgoat650) [reply tests/tools/kopiarunner/kopia_snapshotter.go:371]: you could run it in a container, e.g. go container

File: tests/robustness/engine/engine.go (hunk 2 above)
Comment (rahulchheda) [reply tests/robustness/engine/engine.go:342]: `e.serverCmd` is the process that is the Kopia API Server, to kill this process, we need to send `SIGTERM` signal and run it. ref: https://medium.com/@ganeshmaharaj/clean-exit-of-golangs-exec-command-897832ac3fa5.
And this seems to work, as when I run this suite multiple times, I never get the `51515 port not available`, which use to happen when I didn't add this block of code.
I will check once more, and confirm you.

**Comment by pavannd1** [comment]
@julio-lopez looks like the reviews are addressed and CI is passing. PTAL 🙏🏼

File: tests/robustness/engine/engine.go (hunk 2 above)
Comment (julio-lopez) [reply tests/robustness/engine/engine.go]: What happens when `e.serverCmd.Run()` is called above?

We can address this in a separate PR.

**Review by julio-lopez — COMMENTED** [review]
Let's merge this PR as is and address the server cleanup issue separately.

**Pull request merged**